File: CSE111-Assignment1/rpnc.js
/*
 * Node.js / JavaScript Reverse Polish Notation (RPN) Calculator
 *
 * Accepts an single argument in RPN, evaluates it and prints the answer to stdout.
 * 
 */

//nodejs.org/en/knowledge/command-line/how-to-parse-command-line-arguments
//https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function
//www.w3schools.com/jsref/jsref_length_string.asp
//https://www.w3schools.com/jsref/jsref_string.asp
//https://stackoverflow.com/questions/3586775/what-is-the-correct-way-to-check-for-string-equality-in-javascript
//stackoverflow.com/questions/8935632/check-if-character-is-number/
//https://www.w3schools.com/jsref/jsref_concat_string.asp



let argument = String(process.argv.slice(2));
let n = argument.length; 
let stack = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]; 
let next = 0;
let op2 = 0.0;

function pop(){
	return stack[--next];
}

function push(operand){
	stack[next++] = operand;
}

for(let i = 0; i < n; i++){
	let irrelevant = false;
	if(argument[i] === '+'){
		push(pop()+pop());
	}
	else if(argument[i] === '-'){
		op2 = pop();
		push(pop()-op2);
	} 
	else if(argument[i] === '*'){
		push(pop()*pop());
	} 
	else if(argument[i] === '/'){
		op2 = pop();
		push(pop()/op2);
	} 
	else if(argument[i] === '^'){
		op2 = pop();
		push(Math.pow(pop(), op2));
	}
	else if(!(argument[i] === ' ')){
		if(i != 0){
			if(argument[i-1] >= '0' && argument[i-1] <= '9'){
				irrelevant = true;
			}
		}
		if(irrelevant == false){
			if(i != n - 1){
				if(argument[i+1] >= '0' && argument[i+1] <= '9'){
					if(argument[i+2] >= '0' && argument[i+2] <= '9'){
						push(parseInt(argument[i].concat(argument[i+1]).concat([i+2])));
					}
					else{
						push(parseInt(argument[i].concat(argument[i+1])));
					}
				}
				else{
					push(parseInt(argument[i]));
				}
			}
			else{
				push(parseInt(argument[i]));
			}
		}
	} 
}

console.log(stack[0]);
	

	

File: CSE111-Assignment1/rpnc.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

/*
 * Reverse Polish Notation (RPN) Calculator
 *
 * Accepts a single argument in RPN, evaluates it and prints the answer to stdout.
 * 
 * Returns -1 on error, 0 otherwise.
 */

//borrowed from class

static double stack[16];
static int next = 0;


static double pop(){
	return stack[--next];
}

static void push(double operand){
	stack[next++] = operand;
}

int main(int argc, char *argv[]) {
	double op2;
	char *oper = strtok(argv[1], " ");
	while(oper != NULL){
		switch(oper[0]){
		case '+':
			push(pop() + pop());
			break;
		case '-':
			op2 = pop();
			push(pop() - op2);
			break;
		case '*':
			push(pop() * pop());
			break;
		case '/':
			op2 = pop();
			push(pop()/op2);
			break;
		case '^':
			op2 = pop();
			push(pow(pop(),op2));
			break;
		default:
			push(atoi(oper));
			break;
		}
		oper = strtok(NULL, " ");
	}
	printf("%g\n", stack[0]);
}
			

